spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraRDDPartitionGenerator.scala

```diff
@@ -0,0 +1,323 @@
+    val replicas = metadata.getReplicas(Metadata.quote(keyspaceName), range).map(_.getAddress).toSet
+    val dataSize = (tokenFactory.ringFraction(startToken, endToken) * totalDataSize).toLong
+    new TokenRange(startToken, endToken, replicas, dataSize)
+  }
+
+  def fullRange: TokenRange = {
+
+    val replicas = connector
+      .withClusterDo{cluster => cluster.getMetadata}
+      .getAllHosts
+      .map(_.getAddress)
+      .toSet
+
+    new TokenRange(
+      tokenFactory.minToken,
+      tokenFactory.maxToken,
+      replicas,
+      0)
+  }
+
+  private def describeRing: Seq[TokenRange] = {
+    connector.withClusterDo { cluster =>
+      val metadata = cluster.getMetadata
+      for (tr <- metadata.getTokenRanges.toSeq) yield tokenRange(tr, metadata)
+    }
+  }
+
+  private def splitsOf(
+      tokenRanges: Iterable[TokenRange],
+      splitter: TokenRangeSplitter[V, T]): Iterable[TokenRange] = {
+
+    val parTokenRanges = tokenRanges.par
+    parTokenRanges.tasksupport = new ForkJoinTaskSupport(CassandraRDDPartitionGenerator.pool)
+    (for (tokenRange <- parTokenRanges;
+          split <- splitter.split(tokenRange, splitSize)) yield split).seq
+  }
+
+  private def splitToCqlClause(range: TokenRange): Iterable[CqlTokenRange] = {
+    val startToken = range.start.value
+    val endToken = range.end.value
+    val wrapAround = range.isWrapAround
+    val pk = tableDef.partitionKey.map(_.columnName).map(quote).mkString(", ")
+
+    if (range.end == tokenFactory.minToken)
+      List(CqlTokenRange(GreaterThan(pk), startToken))
+    else if (range.start == tokenFactory.minToken)
+      List(CqlTokenRange(LessThanOrEquals(pk), endToken))
+    else if (!range.isWrapAround)
+      List(CqlTokenRange(GreaterThanAndLessThanOrEquals(pk), startToken, endToken))
+    else
+      List(
+        CqlTokenRange(GreaterThan(pk), startToken),
+        CqlTokenRange(LessThanOrEquals(pk), endToken))
+  }
+
+  private def createTokenRangeSplitter: TokenRangeSplitter[V, T] = {
+    tokenFactory.asInstanceOf[TokenFactory[_, _]] match {
+      case TokenFactory.RandomPartitionerTokenFactory =>
+        new RandomPartitionerTokenRangeSplitter(totalDataSize).asInstanceOf[TokenRangeSplitter[V, T]]
+      case TokenFactory.Murmur3TokenFactory =>
+        new Murmur3PartitionerTokenRangeSplitter(totalDataSize).asInstanceOf[TokenRangeSplitter[V, T]]
+      case _ =>
+        throw new UnsupportedOperationException(s"Unsupported TokenFactory $tokenFactory")
+    }
+  }
+
+  lazy val partitions = partitionsAndRanges.map(_._1)
+  lazy val ranges = partitionsAndRanges.map(x =>
+    (x._2, x._3.map( range =>
+      (range.start, range.end, range.isWrapAround))))
+
+
+  /**
+    * Generates a list of the CassandraPartitions to be used in the TableScanRDD and a list
+    * of the TokenRanges used to craft those partitions along with their indexes.
+    */
+  lazy val partitionsAndRanges: Seq[(CassandraPartition, Int, Seq[TokenRange])] = {
```

Comment:
Why `Seq[(CassandraPartition, Int, Seq[TokenRange])]`? the second component is index which is also present in `CassandraPartition`